app/src/main/java/pl/pwr/zpi/bcycle/mobile/ui/settings/SettingsFragment.kt

```diff
@@ -21,8 +49,300 @@ class SettingsFragment : Fragment() {
+
+        allControls = listOf(cameraBt,passwordPT,repeatPasswordPT,saveBt,namePT,galleryBt,cameraBt,cancelBt)
+    }
+
+    private fun saveChanges() {
+
+        showSpinnerAndDisableControls()
+
+        var newName : String? = null
+        var newPassword : String?  = null
+        var newAvatar : Uri? = null
+
+        if(isNameEdited()){
+            newName = namePT.content()
+        }
+
+        if(isPasswordEditedOk()){
+            newPassword = passwordPT.content()
+        }
+
+        if(isImageEdited()) {
+            newAvatar = mImageUri
+        }
+
+        val handler = Handler()
+        handler.postDelayed(Runnable {
+            changeData(newName,newPassword,newAvatar)
+            // Actions to do after 3 seconds
+        }, 3000)
+
+    }
+
+    fun changeData(name: String?, password: String?, imageUri : Uri?) {
+
+        val profileBuilder = UserProfileChangeRequest.Builder()
+
+        if(password!=null) {
+            currentUser!!.updatePassword(password)
+
+        }
+
+        if(name!=null){
+            profileBuilder.setDisplayName(name)
+        }
+
+        if(imageUri!=null) {
+            uploadImageToFirebase(imageUri)
+            val avatar: Uri? = if(uploadImageToFirebase(imageUri)) {
+                Uri.parse("${FirebaseStorage.getInstance().reference.root}/${currentUser!!.uid}.png")
+            } else {
+                defaultAvatarUri
+            }
+            profileBuilder.setPhotoUri(avatar)
+        }
+
+        if(name!=null || imageUri!=null){
+            val profile = profileBuilder.build()
+
+            currentUser!!.updateProfile(profile).addOnCompleteListener{
+                callback.onDataChanged()
+                hideSpinnerAndEnableControls()
+                activity?.onBackPressed()
+                Toast.makeText(activity,getString(R.string.data_changed_prompt),Toast.LENGTH_SHORT).show()
+
+            }.addOnFailureListener{
+                onFailure()
+            }
+        } else {
+            hideSpinnerAndEnableControls()
+            activity?.onBackPressed()
+            Toast.makeText(activity,getString(R.string.data_changed_prompt),Toast.LENGTH_SHORT).show()
+        }
+
+    }
+
+    private fun onFailure(){
+
+        hideSpinnerAndEnableControls()
+        Toast.makeText(activity,getString(R.string.changed_data_error),Toast.LENGTH_SHORT).show()
+    }
+
+    private fun uploadImageToFirebase(mImageUri:Uri?) :Boolean {
+        if(mImageUri!=null) {
+            val imageRef = FirebaseStorage.getInstance().reference.child(currentUser!!.uid+".png")
+            imageRef.putFile(mImageUri)
+            return true
+        }
+        return false
+    }
+
+    private fun isDataEdited(): Boolean =
+                isNameEdited()  || isImageEdited() || isPasswordEditedOk()
+
+    private fun isNameEdited(): Boolean = namePT.content().isNotEmpty()
+
+    private fun isImageEdited(): Boolean = mImageUri!=null
+
+    private fun isPasswordEditedOk(): Boolean {
```

Comment:
This code (passwords and images) seems to be duplicated in `RegisterActivity.kt`. Perhaps we could move it to a separate class to avoid duplicated code and having to change things in more than one place?

Reply:
zostawiłabym to tak jednak osobno, bo jednak troszkę inaczej to działa, tutaj musi być jedno uzupełnione a w rejestracji jest jednak trochę inaczej